README.md
# hdcquery

CLI tool to query data from [hub.docker.com](https://hub.docker.com).

## Demo

<div style="text-align: center">

[![Demo for hdcquery](https://asciinema.org/a/GID4GzyH6HYYzvmlEkqF6eiTK.svg)](https://asciinema.org/a/GID4GzyH6HYYzvmlEkqF6eiTK)

</div>

## Changelog

Please see the [CHANGELOG](./CHANGELOG.md).

## Usage

### Search for repositories

With `hdcquery search` you can search for repositories:

```console
$ hdcquery search redis
     IMAGE                          DESCRIPTION                                               LAST UPDATE  PULLS  STARS
   1 redis                          Redis is an open source key-value store that …            8 hours ago   10M+   8854
   2 rediscommander/redis-commander Alpine image for redis-commander - Redis …                 9 days ago   10M+     48
   3 redislabs/redisearch           Redis With the RedisSearch module pre-loaded. See …        3 days ago    1M+     29
   4 redislabs/redisinsight         RedisInsight - The GUI for Redis                          16 days ago    1M+     15
   5 redislabs/redismod             An automated build of redismod - latest Redis …            3 days ago  500K+      7
   6 redislabs/rejson               RedisJSON - Enhanced JSON data type processing …          30 days ago  500K+     22
   […]
```

After every page you can type a number to show more details of one of the results, or go to the next page of the search.

    [Found 14432 results] <ENTER>: more results | Number and <ENTER>: image details >

You can limit the numbers of results with the `-l` / `--limit` option.

<details>
<summary>Options for <code>search</code>.</summary>

```console
$ hdcquery --help search
Usage: hdcquery search [OPTIONS]

Positional arguments:
  terms

Optional arguments:
  -l, --limit LIMIT  Limit the number of results
  -s, --search-url SEARCH-URL
                     URL to send search requests
```
</details>


### Show repository details

`hdcquery show` can show the metadata and the description of a repository:

```console
$ hdcquery show redis
Namespace: library
Name: redis
Description: Redis is an open source key-value store that functions as a data structure server.
Starts: 8853
Pulls: 2147483647
Automated: no
Last updated: 2020-12-11 22:02 UTC (10 hours ago)

----

# Quick reference

-	**Maintained by**:
	[the Docker Community](https://github.com/docker-library/redis)
[…]
```

You can skip the metadata and show only the description of the repository with the `-o` / `--only-description` option.

By default, the output of the `show` command is sent to a pager (like `less(1)`). You can use your own pager setting the `HDC_PAGER` variable.

For example, to enable syntax highlighting with [Pygments](https://pygments.org/):

```
$ export HDC_PAGER='sh -c "pygmentize -l md | less -FR"'
```

The pager receives two environment variables:

* `HDCQUERY_REPOSITORY`: the name of the repository (like `library/redis`).
* `HDCQUERY_VERSION`: the version of hdcquery.

<details>
<summary>Options for <code>show</code>.</summary>

```
$ hdcquery --help show
Usage: hdcquery show [OPTIONS]

Positional arguments:
  repositories

Optional arguments:
  -o, --only-description  Only show full description
```
</details>

### List image tags

`hdcquery tags` list tags available in a repository.

```console
$ hdcquery tags redis
- 421 results for redis
SIZE       OS       ARCH   LAST PUSHED     NAME
36.7 MiB   linux    386    10 hours ago    latest
31.5 MiB   linux    arm    10 hours ago    latest
33.8 MiB   linux    arm    10 hours ago    latest
35.1 MiB   linux    arm64  10 hours ago    latest
36.4 MiB   linux    amd64  10 hours ago    latest
35.1 MiB   linux    s390x  10 hours ago    latest
35.1 MiB   linux    arm64  10 hours ago    buster
[…]
```

The option `-l` / `--limit` limits the number of results to show.

The option `-a` / `--architecture` filters the results to the given architecture (`amd64`, `arm`, `386`, …).

The option `-o` / `--operating-system` filters the results to the given operating system (`linux` or `windows`).

The option `-c` / `--current-machine` filters the results to match the operating system and the architecture of the machine where the tool is executed.

Finally, the option `-d` / `--digest` shows the image digest for every tag:

```console
$ hdcquery tags -l4 -dc redis
- 421 results for redis
SIZE       OS       ARCH   LAST PUSHED    DIGEST                                                                    NAME
36.4 MiB   linux    amd64  10 hours ago   sha256:466da50d1e0ba009816a4b507a9b526a34169e026e967f304679b1553cbca66c   latest
36.4 MiB   linux    amd64  10 hours ago   sha256:466da50d1e0ba009816a4b507a9b526a34169e026e967f304679b1553cbca66c   buster
10.1 MiB   linux    amd64  10 hours ago   sha256:4920debee18fad71841ce101a7867743ff8fe7d47e6191b750c3edcfffc1cb18   alpine3.12
10.1 MiB   linux    amd64  10 hours ago   sha256:4920debee18fad71841ce101a7867743ff8fe7d47e6191b750c3edcfffc1cb18   alpine
```

<details>
<summary>Options for <code>tags</code>.</summary>

```console
$ hdcquery --help tags
Usage: hdcquery tags [OPTIONS]

Positional arguments:
  repositories

Optional arguments:
  -l, --limit LIMIT      Limit the number of results (default: 30)
  -d, --digest           Show image digest
  -a, --architecture ARCHITECTURE
                         Filter by architecture
  -o, --operating-system OPERATING-SYSTEM
                         Filter by operating system
  -c, --current-machine  Filter by operating system and architecture of this machine
```
</details>

## Installation

### Pre-built binaries

Every release includes binaries for Linux, macOS, and Windows.

You can download them in the [*Releases* page](https://github.com/ayosec/hdcquery/releases).

The binaries are built in a [GitHub Actions workflow](./.github/workflows/release.yml), so they are limited to the platforms supported by GitHub.

### Using Docker

You can run the tool with Docker, using the `ayosec/hdcquery` image.

```console
$ docker run --rm ayosec/hdcquery --help
```

Like the pre-built binaries, the Docker image is built in another [GitHub Actions workflow](./.github/workflows/docker.yml) for every new release.

### From sources

To build the tool from sources:

1. [Install the Rust toolchain](https://www.rust-lang.org/tools/install).

2. Use `cargo install`:

    ```console
    $ cargo install --git https://github.com/ayosec/hdcquery.git
    ```

## Similar tools

* [amalfra/docker-hub](https://github.com/amalfra/docker-hub)
* [containers/skopeo](https://github.com/containers/skopeo)


CHANGELOG.md
# Changelog

## 0.1.1

* Parse `$HDC_PAGER` according to parsing rules of Unix shell.

## 0.1

* Added `search` command.
* Added `show` command.
* Added `tags` command.


src/show.rs
//! Implementation of the 'show' command
//!
//! The pager will receive the following environment variables:
//!
//! * `HDCQUERY_REPOSITORY`: name of the repository, like `"library/rust"`.
//! * `HDCQUERY_VERSION`: version of hdcquery.

use crate::hubapi::Repository;
use crate::langext::DurationExt;
use crate::options::ShowOptions;

const REPOSITORY_URL: &str = "https://hub.docker.com/v2/repositories/";

pub async fn run(options: ShowOptions) -> anyhow::Result<()> {
    if options.repositories.is_empty() {
        eprintln!("No repositories");
        return Ok(());
    }

    for repository in &options.repositories {
        let repository = get_repository(&repository).await?;
        if options.only_description {
            if let Some(full_description) = repository.full_description {
                println!("{}", full_description);
            }
        } else {
            show_repository(&repository).await?
        }
    }

    Ok(())
}

pub async fn show_repository_by_slug(slug: &str) -> anyhow::Result<()> {
    show_repository(&get_repository(slug).await?).await
}

pub async fn show_repository(repository: &Repository) -> anyhow::Result<()> {
    let pager_cmd;
    let mut pager_stdin;
    let mut io_stdout;

    let output: &mut dyn std::io::Write;

    macro_rules! option_field {
        ($field:ident, $label:literal) => {
            option_field!($field, $label, $field)
        };

        ($field:ident, $label:literal, $render:expr) => {
            if let Some($field) = &repository.$field {
                writeln!(output, concat!($label, ": {}"), $render)?
            }
        };
    }

    let slug = match (&repository.namespace, &repository.name) {
        (Some(ns), Some(n)) => Some(format!("{}/{}", ns, n)),
        _ => None,
    };

    if let Some(mut cmd) = crate::pager::command(slug.as_deref()) {
        let mut child = cmd.spawn()?;
        pager_stdin = child.stdin.take();
        pager_cmd = Some(child);

        output = pager_stdin.as_mut().expect("stdin expected for pager");
    } else {
        pager_cmd = None;
        pager_stdin = None;
        io_stdout = std::io::stdout();
        output = &mut io_stdout;
    }

    option_field!(namespace, "Namespace");
    option_field!(name, "Name");
    option_field!(description, "Description");
    option_field!(star_count, "Starts");
    option_field!(pull_count, "Pulls");

    option_field!(
        is_automated,
        "Automated",
        if *is_automated { "yes" } else { "no" }
    );

    if let Some(last_updated) = &repository.last_updated {
        writeln!(
            output,
            "Last updated: {} ({})",
            last_updated.format("%F %R %Z"),
            last_updated.to_human()
        )?;
    }

    if let Some(full_description) = &repository.full_description {
        writeln!(output, "\n----\n\n{}\n\n----", full_description)?;
    }

    // Close the pager's input
    drop(pager_stdin);
    let _ = pager_cmd.map(|mut c| c.wait());

    Ok(())
}

/// Download the repository data.
///
/// If `slug` does not contain a '/', it will be prepended with "library/".
/// This is the expected format by hub.docker.com
async fn get_repository(slug: &str) -> Result<Repository, reqwest::Error> {
    let full_url = if slug.contains('/') {
        format!("{}{}/", REPOSITORY_URL, slug)
    } else {
        format!("{}library/{}/", REPOSITORY_URL, slug)
    };

    let http_client = crate::hubapi::http_client()?;

    http_client
        .get(&full_url)
        .header("Accept", "application/json")
        .send()
        .await?
        .json()
        .await
}

#[test]
fn show_rust_repository() {
    use assert_cmd::prelude::*;
    use std::process::Command;

    let process = {
        let mut cmd = Command::cargo_bin(env!("CARGO_PKG_NAME")).unwrap();
        cmd.args(&["show", "rustlang/rust"]);
        cmd.unwrap()
    };

    let stdout = std::str::from_utf8(&process.stdout).unwrap();

    assert!(stdout.contains("Namespace: rustlang"));
    assert!(stdout.contains("Name: rust"));
    assert!(stdout.contains("---"));
    assert!(process.status.code() == Some(0));
}


src/pager.rs
//! Run an external program as a pager (like "less")

use std::env;
use std::process::{Command, Stdio};

pub const PAGER_ENV: &str = "HDC_PAGER";

const DEFAULT_PAGER: &str = if cfg!(windows) { "more" } else { "pager -F" };

pub fn command(repository: Option<&str>) -> Option<Command> {
    if !atty::is(atty::Stream::Stdout) {
        return None;
    }

    let pager_var = env::var(PAGER_ENV);
    let pager_args = pager_var.as_deref().unwrap_or(DEFAULT_PAGER);

    let mut pager_args = match shell_words::split(pager_args) {
        Ok(words) => words.into_iter(),
        Err(e) => {
            eprintln!("Failed to parse $HDC_PAGER: {:?}", e);
            return None;
        }
    };

    let mut cmd = Command::new(pager_args.next()?);

    for arg in pager_args {
        cmd.arg(arg);
    }

    cmd.stdin(Stdio::piped());
    cmd.env("HDCQUERY_VERSION", env!("CARGO_PKG_VERSION"));
    repository.map(|r| cmd.env("HDCQUERY_REPOSITORY", r));

    Some(cmd)
}


src/hubapi.rs
//! Types and common functions to use API from hub.docker.com

use chrono::prelude::*;
use std::time::Duration;

pub fn http_client() -> Result<reqwest::Client, reqwest::Error> {
    reqwest::Client::builder()
        .user_agent(concat!("dhrb/", env!("CARGO_PKG_VERSION")))
        .timeout(Duration::from_secs(30))
        .connect_timeout(Duration::from_secs(10))
        .build()
}

#[derive(serde::Deserialize, Debug)]
pub struct Summary {
    pub name: String,
    pub slug: String,
    pub updated_at: DateTime<Utc>,
    pub pull_count: Option<String>,
    pub star_count: Option<usize>,
    pub short_description: Option<String>,
}

#[derive(serde::Deserialize, Debug)]
pub struct Repository {
    pub namespace: Option<String>,
    pub name: Option<String>,
    pub description: Option<String>,
    pub star_count: Option<usize>,
    pub pull_count: Option<usize>,
    pub full_description: Option<String>,
    pub is_automated: Option<bool>,
    pub last_updated: Option<DateTime<Utc>>,
}

#[derive(serde::Deserialize, Debug)]
pub struct Tag {
    pub name: String,
    pub images: Vec<Image>,
    pub last_updated: Option<DateTime<Utc>>,
}

#[derive(serde::Deserialize, Debug)]
pub struct Image {
    pub architecture: String,
    pub os: String,
    pub size: u64,
    pub digest: Option<String>,
}


src/search.rs
//! Implementation of the 'search' command

use std::io::Write;
use std::str::FromStr;

use crate::hubapi::Summary;
use crate::langext::DurationExt;
use crate::options::SearchOptions;

use tokio::io::AsyncBufReadExt;
use tokio::sync::mpsc;

const DEFAULT_SEARCH_URL: &str = "https://hub.docker.com/api/content/v1/products/search";

const PAGES_QUEUE_SIZE: usize = 2;

#[derive(serde::Deserialize, Debug)]
struct SearchResult {
    count: usize,
    summaries: Option<Vec<Summary>>,
}

pub async fn run(options: SearchOptions) -> anyhow::Result<()> {
    let (term_width, term_height) = match terminal_size::terminal_size() {
        Some((terminal_size::Width(w), terminal_size::Height(h))) => (w as usize, h as usize),
        None => (80, 25),
    };

    let show_prompt = atty::is(atty::Stream::Stdout) && atty::is(atty::Stream::Stdin);

    let description_width = term_width - /* column widths */ 5 - 31 - 18 - 7 - 7;

    let limit = options.limit.unwrap_or(usize::MAX);
    let terms = options.terms.join(" ");
    let search_url = options.search_url;

    macro_rules! row {
        ($($values:tt)*) => {
            println!(
                "{:4} {:30.30} {:dw$.dw$} {:>17.17} {:>6.6} {:>6.6}",
                $($values)*,
                dw = description_width
            )
        }
    }

    let (pages_tx, mut pages_rx) = mpsc::channel(PAGES_QUEUE_SIZE);
    tokio::spawn(async move {
        pages_queue(pages_tx, search_url, terms, term_height - 2)
            .await
            .unwrap()
    });

    let mut repositories_found = vec![];

    let mut stdin = tokio::io::BufReader::new(tokio::io::stdin()).lines();

    'main: while let Some(results) = pages_rx.recv().await {
        let total_rows = results.count;
        let summaries = match results.summaries {
            Some(s) if !s.is_empty() => s,
            _ => break,
        };

        // Print current page.

        if show_prompt || repositories_found.is_empty() {
            row!("", "IMAGE", "DESCRIPTION", "LAST UPDATE", "PULLS", "STARS");
        }

        for summary in summaries {
            row!(
                repositories_found.len() + 1,
                summary.slug,
                first_line(summary.short_description.as_ref(), Some(description_width)),
                summary.updated_at.to_human(),
                first_line(summary.pull_count.as_ref(), None),
                summary.star_count.unwrap_or(0)
            );

            repositories_found.push(summary);

            if repositories_found.len() >= limit {
                break;
            }
        }

        // Get input from user.

        loop {
            if !show_prompt {
                break;
            }

            print!(
                "[Found {} results] <ENTER>: more results | Number and <ENTER>: image details > ",
                total_rows
            );

            std::io::stdout().flush()?;

            let line = stdin.next_line().await?.unwrap_or_default();
            let number: usize = match line.trim() {
                "" => break,

                l => match usize::from_str(l) {
                    Ok(n) => n,
                    Err(e) => {
                        eprintln!("{:?}: {}", l, e);
                        continue;
                    }
                },
            };

            let repository = match repositories_found.get(number.wrapping_sub(1)) {
                Some(i) => i,

                None => {
                    eprintln!("Image not found");
                    continue;
                }
            };

            crate::show::show_repository_by_slug(&repository.slug).await?;
            break 'main;
        }

        if repositories_found.len() >= limit {
            break 'main;
        }
    }

    Ok(())
}

/// Extract the first line of `value`.
///
/// If `value` is `None`, returns an empty string.
///
/// If `width` is defined, the text will be wrapped to it.
fn first_line<T>(value: Option<T>, width: Option<usize>) -> String
where
    T: std::fmt::Display,
{
    let value = match value {
        Some(v) => v.to_string(),
        None => return String::new(),
    };

    match (width, value.split('\n').next()) {
        (_, None) => String::new(),

        (None, Some(v)) => v.into(),

        (Some(width), Some(value)) => {
            let mut lines = textwrap::wrap(value, width - 2).into_iter();
            let first_line = lines.next().unwrap_or_default();

            if lines.next().is_some() {
                // Append '…' to the result if there are more than one line.
                format!("{} …", first_line)
            } else {
                first_line.to_string()
            }
        }
    }
}

/// Background task to download pages when the program is waiting for user
/// input.
///
/// The queue is bounde to PAGES_QUEUE_SIZE, so we will not download a lot of
/// unneeded pages.
async fn pages_queue(
    mut pages_tx: mpsc::Sender<SearchResult>,
    search_url: Option<String>,
    terms: String,
    per_page: usize,
) -> anyhow::Result<()> {
    let per_page = format!("{}", per_page);

    let search_url = search_url
        .as_ref()
        .map(|s| &s[..])
        .unwrap_or(DEFAULT_SEARCH_URL);

    let mut num_page: usize = 1;

    let http_client = crate::hubapi::http_client()?;

    loop {
        let np_str = num_page.to_string();

        let result = http_client
            .get(search_url)
            .header("Accept", "application/json")
            .header("Search-Version", "v3")
            .query(&[
                ("type", "image"),
                ("q", terms.as_str()),
                ("page_size", per_page.as_str()),
                ("page", np_str.as_str()),
            ])
            .send()
            .await?
            .json()
            .await?;

        if pages_tx.send(result).await.is_err() {
            // Receiver is closed
            return Ok(());
        }

        num_page += 1;
    }
}

#[cfg(target_os = "linux")]
#[test]
fn search_apache_repository() {
    use assert_cmd::cargo::CommandCargoExt;
    use rexpect::process::wait::WaitStatus::Exited;
    use rexpect::session::spawn_command;
    use std::process::Command;

    let mut cmd = Command::cargo_bin(env!("CARGO_PKG_NAME")).unwrap();
    cmd.env(crate::pager::PAGER_ENV, "cat");
    cmd.args(&["search", "apache"]);
    let mut cmd = spawn_command(cmd, Some(60_000)).unwrap();

    // Find "httpd" repository

    let (_, line) = cmd.exp_regex("\n +[0-9]+ +httpd ").unwrap();
    let repostiory_number = line.split_whitespace().next().unwrap();
    cmd.send_line(repostiory_number).unwrap();

    // Show repository details

    cmd.exp_string("Namespace: library").unwrap();
    cmd.exp_string("Name: httpd").unwrap();
    cmd.exp_string("# Quick reference").unwrap();

    assert!(matches!(cmd.process.wait(), Ok(Exited(_, 0))));
}


src/langext.rs
use chrono::prelude::*;

pub trait DurationExt {
    fn to_human(&self) -> String;
}

impl DurationExt for DateTime<chrono::Utc> {
    fn to_human(&self) -> String {
        let time_ago = Utc::now() - *self;

        if time_ago.num_seconds() < 90 {
            return "just now".to_string();
        }

        if let n @ 0..=120 = time_ago.num_minutes() {
            return format!("{} minutes ago", n);
        }

        if let n @ 0..=72 = time_ago.num_hours() {
            return format!("{} hours ago", n);
        }

        if let n @ 0..=30 = time_ago.num_days() {
            return format!("{} days ago", n);
        }

        if let n @ 0..=20 = time_ago.num_weeks() {
            return format!("{} weeks ago", n);
        }

        self.format("%F").to_string()
    }
}

#[test]
fn compute_durations() {
    macro_rules! check {
        ($num:literal $unit:ident => $expect:expr) => {
            let time = Utc::now() - chrono::Duration::$unit($num);
            assert_eq!(time.to_human(), $expect.to_string());
        };
    }

    check!(10 seconds => "just now");
    check!(30 minutes => "30 minutes ago");
    check!(8 hours => "8 hours ago");
    check!(10 days => "10 days ago");
    check!(15 weeks => "15 weeks ago");

    let long_ago = DateTime::parse_from_rfc3339("2000-01-02T00:00:00Z").unwrap();
    assert_eq!(
        long_ago.with_timezone(&Utc).to_human(),
        "2000-01-02".to_string()
    );
}


src/options.rs
#[derive(Debug, gumdrop::Options)]
pub struct Options {
    #[options(help = "print help message")]
    pub help: bool,

    #[options(command)]
    pub command: Option<Command>,
}

#[derive(Debug, gumdrop::Options)]
pub enum Command {
    #[options(help = "search for repositories")]
    Search(SearchOptions),

    #[options(help = "show repository details")]
    Show(ShowOptions),

    #[options(help = "show repository tags")]
    Tags(TagsOptions),
}

#[derive(Debug, gumdrop::Options)]
pub struct SearchOptions {
    #[options(help = "Limit the number of results")]
    pub limit: Option<usize>,

    #[options(help = "URL to send search requests")]
    pub search_url: Option<String>,

    #[options(free)]
    pub terms: Vec<String>,
}

#[derive(Debug, gumdrop::Options)]
pub struct ShowOptions {
    #[options(help = "Only show full description")]
    pub only_description: bool,

    #[options(free)]
    pub repositories: Vec<String>,
}

#[derive(Debug, gumdrop::Options)]
pub struct TagsOptions {
    #[options(help = "Limit the number of results", default = "30")]
    pub limit: usize,

    #[options(help = "Show image digest")]
    pub digest: bool,

    #[options(help = "Filter by architecture")]
    pub architecture: Option<String>,

    #[options(help = "Filter by operating system")]
    pub operating_system: Option<String>,

    #[options(help = "Filter by operating system and architecture of this machine")]
    pub current_machine: bool,

    #[options(free)]
    pub repositories: Vec<String>,
}


src/tags.rs
//! Implementation of the 'tags' command

use crate::hubapi::Tag;
use crate::langext::DurationExt;
use crate::options::TagsOptions;

const DEFAULT_OS: Option<&str> = if cfg!(target_os = "linux") {
    Some("linux")
} else if cfg!(target_os = "windows") {
    Some("windows")
} else {
    None
};

const DEFAULT_ARCH: Option<&str> = if cfg!(target_arch = "x86_64") {
    Some("amd64")
} else if cfg!(target_arch = "x86") {
    Some("386")
} else if cfg!(target_arch = "arm") {
    Some("arm")
} else if cfg!(target_arch = "aarch64") {
    Some("arm64")
} else {
    None
};

#[derive(serde::Deserialize, Debug)]
struct Response {
    count: usize,
    results: Vec<Tag>,
}

pub async fn run(options: TagsOptions) -> anyhow::Result<()> {
    macro_rules! row {
        ($size:expr, $os:expr, $arch:expr, $push:expr, $digest:expr, $name:expr) => {
            if options.digest {
                println!(
                    "{:10} {:8.8} {:6.6} {:14} {:73} {}",
                    $size, $os, $arch, $push, $digest, $name
                )
            } else {
                println!(
                    "{:10} {:8.8} {:6.6} {:15} {}",
                    $size, $os, $arch, $push, $name
                )
            }
        };
    }

    let filter_os;
    let filter_arch;

    if options.current_machine {
        filter_os = DEFAULT_OS;
        filter_arch = DEFAULT_ARCH;
    } else {
        filter_os = options.operating_system.as_deref();
        filter_arch = options.architecture.as_deref();
    }

    let http_client = crate::hubapi::http_client()?;

    let mut pending = options.limit;

    for repository in &options.repositories {
        let slug_prefix = if repository.contains('/') {
            ""
        } else {
            "library/"
        };

        'repository: for page in 1.. {
            let url = format!(
                "https://hub.docker.com/v2/repositories/{}{}/tags/?page={}&page_size={}",
                slug_prefix,
                repository,
                page,
                options.limit.min(50)
            );

            let response: Response = http_client
                .get(&url)
                .header("Accept", "application/json")
                .send()
                .await?
                .json()
                .await?;

            if page == 1 {
                println!("- {} results for {}", response.count, repository);

                row!("SIZE", "OS", "ARCH", "LAST PUSHED", "DIGEST", "NAME");
            }

            if response.results.is_empty() {
                break;
            }

            for result in response.results {
                let last_updated = result
                    .last_updated
                    .map(|lp| lp.to_human())
                    .unwrap_or_default();

                for image in &result.images {
                    if pending == 0 {
                        break 'repository;
                    }

                    if filter_os.is_some() && filter_os != Some(image.os.as_str()) {
                        continue;
                    }

                    if filter_arch.is_some() && filter_arch != Some(image.architecture.as_str()) {
                        continue;
                    }

                    pending -= 1;

                    row!(
                        bytesize::to_string(image.size, true),
                        image.os,
                        image.architecture,
                        last_updated,
                        image.digest.as_deref().unwrap_or_default(),
                        result.name
                    );
                }
            }
        }
    }

    Ok(())
}
